### PR TITLE
Feature/share dictionaries

### PR DESCRIPTION
**Pull request by VargaJohanna**

### CODE REVIEW

File: app/src/main/java/com/vocabulary/myvocabulary/ui/dictionaries/ShareDictionaryViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    override fun onCleared() {
+        disposables.clear()
+        super.onCleared()
+    }
+
+    fun createDictionary(dictionary: Dictionary) {
+        disposables += Single.fromCallable { dictionaryRepository.createDictionary(dictionary) }
+                .subscribeOn(rxSchedulers.io())
+                .observeOn(rxSchedulers.main())
+                .subscribe { t: Long ->
```
Comment (jbarat) [inline app/src/main/java/com/vocabulary/myvocabulary/ui/dictionaries/ShareDictionaryViewModel.kt]: name it properly or remove it and use `it`

File: app/src/main/java/com/vocabulary/myvocabulary/ui/dictionaries/ShareDictionaryViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+                    importedDictionaryDetails.postValue(Event(DictionaryDetails(t, dictionary.dictionaryName)))
+                }
+    }
+
+    fun fetchCsvUri(): Uri? = shareDictionaryRepository.getCsvData()
+
+    fun parseDataAndCreateWords(dictionaryId: Long, context: Context) {
+        if(shareDictionaryRepository.getCsvData() != null) {
```
Comment (jbarat) [inline app/src/main/java/com/vocabulary/myvocabulary/ui/dictionaries/ShareDictionaryViewModel.kt]: Change this to use shareDictionaryRepository.getCsvData()?let{}. Then you don't need to double bang it later.

File: app/src/main/java/com/vocabulary/myvocabulary/ui/dictionaries/ShareDictionaryViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+                    for (csvRecord in csvParser) {
+                        insertWordToDatabase(Word(
+                                containerDictionaryId = dictionaryId,
+                                translation = csvRecord.get(0),
+                                word = csvRecord.get(1),
+                                created = Calendar.getInstance().time
+                        ))
+                    }
+                } catch (ex: Exception) {
+                    ex.printStackTrace()
```
Comment (jbarat) [inline app/src/main/java/com/vocabulary/myvocabulary/ui/dictionaries/ShareDictionaryViewModel.kt:83]: Log it to a crash reporting.

File: app/src/main/java/com/vocabulary/myvocabulary/ui/dictionaries/ShareDictionaryViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        shareDictionaryRepository.setIsImport(isImport)
+    }
+
+    fun saveCsvData(csv: Uri) {
+        shareDictionaryRepository.storeCsvData(csv)
+    }
+
+    private fun writeCsvFile(words: List<Word>, context: Context): File {
+        var fileWriter: FileWriter? = null
+        var csvPrinter: CSVPrinter? = null
```
Comment (jbarat) [inline app/src/main/java/com/vocabulary/myvocabulary/ui/dictionaries/ShareDictionaryViewModel.kt]: There is no reason for `fileWriter` and `csvPrinter` to be a nullable var. Move them into the try.

File: app/src/main/java/com/vocabulary/myvocabulary/ui/dictionaries/ShareDictionaryViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+                        word.word
+                )
+                csvPrinter.printRecord(data)
+            }
+        } catch (e: Exception) {
+            e.printStackTrace()
+            Log.e("ERROR", e.message)
+        } finally {
+            try {
+                fileWriter!!.flush()
```
Comment (jbarat) [inline app/src/main/java/com/vocabulary/myvocabulary/ui/dictionaries/ShareDictionaryViewModel.kt]: How will this work if the exception happened at the creation of the `fileWriter`?

File: app/src/main/java/com/vocabulary/myvocabulary/ui/dictionaries/ShareDictionaryViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        intent.type = "text/csv"
+        intent.putExtra(
+                Intent.EXTRA_STREAM,
+                FileProvider.getUriForFile(
+                        context,
+                        context.applicationContext.packageName + ".fileprovider",
+                        file)
+        )
+        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
+        startActivity(context, Intent.createChooser(intent, "Share file"), null)
```
Comment (jbarat) [inline app/src/main/java/com/vocabulary/myvocabulary/ui/dictionaries/ShareDictionaryViewModel.kt]: "Share file" should come from the resources.

File: app/src/main/java/com/vocabulary/myvocabulary/repositories/share/ShareDictionaryRepositoryImpl.kt [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+class ShareDictionaryRepositoryImpl : ShareDictionaryRepository {
+    private val _import = BehaviorSubject.createDefault(false)
+    private val import: Observable<Boolean> = _import
+
+    override fun setIsImport(import: Boolean) {
+        _import.onNext(import)
+    }
+
+    override fun getImport(): Observable<Boolean> {
+        return import
```
Comment (jbarat) [inline app/src/main/java/com/vocabulary/myvocabulary/repositories/share/ShareDictionaryRepositoryImpl.kt]: Use an expression body.

File: app/src/main/java/com/vocabulary/myvocabulary/repositories/share/ShareDictionaryRepositoryImpl.kt [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    private var csvData: Uri? = null
+
+    override fun getCsvData(): Uri? {
+        return csvData
+    }
+
+    override fun storeCsvData(data: Uri) {
+        csvData = data
```
Comment (jbarat) [inline app/src/main/java/com/vocabulary/myvocabulary/repositories/share/ShareDictionaryRepositoryImpl.kt:22]: Use an expression body.

File: app/src/main/java/com/vocabulary/myvocabulary/repositories/share/ShareDictionaryRepositoryImpl.kt [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+        _import.onNext(import)
+    }
+
+    override fun getImport(): Observable<Boolean> {
+        return import
+    }
+
+    private var csvData: Uri? = null
+
+    override fun getCsvData(): Uri? {
```
Comment (jbarat) [inline app/src/main/java/com/vocabulary/myvocabulary/repositories/share/ShareDictionaryRepositoryImpl.kt]: Use an expression body.

File: app/src/main/java/com/vocabulary/myvocabulary/repositories/share/ShareDictionaryRepository.kt [hunk 10]
```diff
@@ -0,0 +1,12 @@
+package com.vocabulary.myvocabulary.repositories.share
+
+import android.net.Uri
+import io.reactivex.Observable
+import io.reactivex.Single
+
+interface ShareDictionaryRepository {
+    fun storeCsvData(data: Uri)
+    fun getCsvData(): Uri?
```
Comment (jbarat) [inline app/src/main/java/com/vocabulary/myvocabulary/repositories/share/ShareDictionaryRepository.kt]: This is not getting the Data. It is getting an URI.

**Review by jbarat — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/vocabulary/myvocabulary/repositories/share/ShareDictionaryRepository.kt (hunk 10 above)
Comment (VargaJohanna) [reply app/src/main/java/com/vocabulary/myvocabulary/repositories/share/ShareDictionaryRepository.kt]: Fixed.

File: app/src/main/java/com/vocabulary/myvocabulary/repositories/share/ShareDictionaryRepositoryImpl.kt (hunk 9 above)
Comment (VargaJohanna) [reply app/src/main/java/com/vocabulary/myvocabulary/repositories/share/ShareDictionaryRepositoryImpl.kt]: Done

File: app/src/main/java/com/vocabulary/myvocabulary/repositories/share/ShareDictionaryRepositoryImpl.kt (hunk 7 above)
Comment (VargaJohanna) [reply app/src/main/java/com/vocabulary/myvocabulary/repositories/share/ShareDictionaryRepositoryImpl.kt]: Done

File: app/src/main/java/com/vocabulary/myvocabulary/ui/dictionaries/ShareDictionaryViewModel.kt (hunk 6 above)
Comment (VargaJohanna) [reply app/src/main/java/com/vocabulary/myvocabulary/ui/dictionaries/ShareDictionaryViewModel.kt]: Done

File: app/src/main/java/com/vocabulary/myvocabulary/ui/dictionaries/ShareDictionaryViewModel.kt (hunk 5 above)
Comment (VargaJohanna) [reply app/src/main/java/com/vocabulary/myvocabulary/ui/dictionaries/ShareDictionaryViewModel.kt]: It won't. I removed the finally block and added the close and flush methods to the first try block.

File: app/src/main/java/com/vocabulary/myvocabulary/ui/dictionaries/ShareDictionaryViewModel.kt (hunk 2 above)
Comment (VargaJohanna) [reply app/src/main/java/com/vocabulary/myvocabulary/ui/dictionaries/ShareDictionaryViewModel.kt]: Done

File: app/src/main/java/com/vocabulary/myvocabulary/ui/dictionaries/ShareDictionaryViewModel.kt (hunk 1 above)
Comment (VargaJohanna) [reply app/src/main/java/com/vocabulary/myvocabulary/ui/dictionaries/ShareDictionaryViewModel.kt]: Can't use it, but I gave it a name

**Pull request merged**